### PR TITLE
router: 2: allow routermiddleware in dispatcher

### PR DESCRIPTION
**Pull request by willhug**
Summary: this diff allows the yarpc.Dispatcher to accept a custom
routing table in the yarpc.Config so that callsites may pass in custom
routing logic that aheres to the transport.RouteTable interface

### CODE REVIEW

**Comment by breerly** [comment]
I'd like to keep the term "middleware" reserved for inbound/outbound middleware - let's jam on a better term for this.

File: dispatcher_test.go [hunk 1]
```diff
@@ -34,6 +35,7 @@ import (
 	"github.com/golang/mock/gomock"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
+	"go.uber.org/yarpc/api/middleware/middlewaretest"
```
Comment (willhug) [inline dispatcher_test.go]: whoops

File: dispatcher.go [hunk 2]
```diff
@@ -188,12 +197,18 @@ func (d *Dispatcher) ClientConfig(service string) transport.ClientConfig {
...
+		return d.routerMiddleware.Procedures(d.table)
+	}
 	return d.table.Procedures()
 }
 
 // Choose picks a handler for the given request or returns an error if a
 // handler for this request does not exist.
 func (d *Dispatcher) Choose(ctx context.Context, req *transport.Request) (transport.HandlerSpec, error) {
+	if d.routerMiddleware != nil {
+		return d.routerMiddleware.Choose(ctx, req, d.table)
```
Comment (abhinav) [inline dispatcher.go]: Use the same pattern here that we use for outbounds and handlers.

Create an `ApplyRouterMiddleware` function which wraps a `Router` with a
`RouterMiddleware` and use that.

You may have to make a private version of it that works with RouteTables.

**Comment by willhug** [comment]
We're going to keep this around, but this is a design we can fix post-1.0

**Comment by willhug** [comment]
will revisit later

**Pull request closed without merging**